### ISSUE TITLE
Public C API

### ISSUE THREAD

**Issue by gumb0** [wasmx/fizzy#558]
- [ ] Validate #530 
  - [ ] returning error code/message?
- [ ] Parse #533 
  - [ ] returning error code/message?
- [ ] Instantiate #533
  - [ ] Providing imports
    - [ ] functions #533 #557
    - [ ] globals
    - [ ] memory
    - [ ] table
  - [ ] Providing memory pages limit
  - [ ] returning error code/message?
- [ ] Execute #533 
- [ ] Module inspection
  - [ ] types
  - [ ] imports
  - [ ] function types
  - [ ] table limits
  - [ ] memory limits
  - [ ] global types
  - [ ] exports
  - [ ] start function?
  - [ ] element, data, code?
  - [ ] custom sections?
- [ ] Exports access (find by name)
- [ ] Resolving imports

**Comment by Qix-** [comment]
Can a roadmap item be added for all-around better error handling? The existing API has zero error handling, and though I'm moving to Fizzy from wasm3 due to some convoluted blocking bugs, wasm3 has _much_ better error handling in its API that I'm really missing.

Right now, it's just "welp, it failed" if some of the resulting pointers are NULL - there's nothing to help understand _why_ something failed.

**Comment by axic** [comment]
You are absolutely right. I was integrating fizzy into something else the other day and the lack of error message propagation is annoying. We'll think about a good API for this for the C layer, but if you have some ideas please do not be afraid to share them.

**Comment by Qix-** [comment]
@Axic an extra parameter that accepts a pointer to an error code would be great.

Alternatively, changing the return types to be an error code (or success code of course) and having the pointer results be returned via pointer-to-pointer arguments.

For example:

```c
FizzyModule *fz_mod;

int r = fizzy_parse(
	&fz_mod,
	mod_bytes.data(),
	mod_bytes.size()
);

if (r != 0) {
    fprintf(stderr, "error: failed to parse module: %s\n", fizzy_strerr(r)); 
} else {
    assert(fz_mod != NULL);
    /* fz_mod is valid */
}
```

**Comment by axic** [comment]
I was also considering two other options. I think we want to go beyond error codes to have the capability of returning messages. While error codes work well for many cases, they do not for telling about import resolutions problems (i.e. including namespace, module, etc.)

There are two problems with returning messages though: a) who manages the memory allocated for them; b) can we avoid allocating memory.

In the following option `FizzyError` is just a nice container for a fixed-size buffer. In this case the error message buffer lives on the stack of the caller:
```
FizzyError err;
fizzy_parse(..., &err);
```

In the following option we use a global buffer, which can be queried. The downside is that subsequent calls overwrite it and it is not thread safe.
```
fizzy_parse(...);
... = fizzy_error(); // returns char*
```

I'm sure @chfast and @gumb0 have some better ideas.

**Comment by axic** [comment]
@Qix- another question: should we have a public C++ API, would that be preferable? I'm not saying we'll have it anytime soon, but it is only missing due to lack of time.

**Comment by chfast** [comment]
Something like this seems ok (just a sketch), but happy to review C API design from other libraries.
```cpp
struct FizzyError { int error_code; const char* message; }

FizzyError err;
fizzy_parse(..., &err);
if (err.error_code != 0)
    printf("%s", err.message) && exit(err.error_code);
fizzy_instantiate(..., &err);
if (err.error_code != 0)
    printf("%s", err.message) && exit(err.error_code);
fizzy_free_error(&err);
```

So the `FizzyError` is reusable and optional (you can pass `NULL` meaning I don't care about error.

**Comment by axic** [comment]
Yes, that matches my first idea. Though I was wondering if the struct should be opaque or not.

However, when it is reusable then subsequent users must check if `err.message` is not null and free it.

**Comment by chfast** [comment]
> However, when it is reusable then subsequent users must check if `err.message` is not null and free it.

No. This can be handled inside Fizzy. Only after the last use the manual free is needed.

Alternatively, we can adopt this `Store` think present in many wasm implementation, including wasm-c-api.

**Comment by axic** [comment]
> > However, when it is reusable then subsequent users must check if err.message is not null and free it.
>
> No. This can be handled inside Fizzy. Only after the last use the manual free is needed.

My wording was unclear, by "subsequent users" I meant subsequent calls. So I agree :wink:

**Comment by Qix-** [comment]
The problem about error codes is two-fold:

- They are not (easily) localizable.
- They are not guaranteed to be programmatically switchable (e.g. `if (errcode == FIZZY_ENOSUCHIMPORT) { /* ignore */ }`) even if you use `extern` strings.

This is why many C libraries provide a `const char * strerror(int)` function that translates error codes to messages (e.g. POSIX C, [libuv](http://docs.libuv.org/en/v1.x/errors.html#c.uv_strerror), etc.)

I would recommend doing it this way.

---

In the event you do decide to go with strings, please under no circumstances perform an allocation. This could cause a cascade of failures that makes debugging (especially in production) very difficult (a memory error would cause a memory-related crash, masking the originating Fizzy-related error - or, many error codes could be generated that are ultimately ignored, but a forgotten `free()` causes a memory leak over time, etc.)
